include/runtime-support.h

```diff
@@ -629,6 +631,64 @@ inline ::zeek::ValPtr to_val(const T& t, ::zeek::TypePtr target, const std::stri
     return zeek::compat::ToValPtr(std::move(rval));
 }
 
+/**
+ * Converts Spicy-side struct to a Zeek record value. The result is returned
+ * with a ref count +1.
+ */
+template<typename T, typename std::enable_if_t<std::is_base_of<::hilti::rt::trait::isStruct, T>::value>*>
+inline ::zeek::ValPtr to_val(const T& t, ::zeek::TypePtr target, const std::string& location) {
+    if ( target->Tag() != ::zeek::TYPE_RECORD )
+        throw TypeMismatch("struct", target, location);
+
+    auto rtype = target->AsRecordType();
+
+    auto rval = std::make_unique<::zeek::RecordVal>(zeek::compat::ToValCtorType(rtype));
+    int idx = 0;
+
+    auto numFields = rtype->NumFields();
+
+    t.__visit([&](const auto& name, const auto& val) {
+        if ( idx >= numFields )
+            throw TypeMismatch(hilti::rt::fmt("no matching record field for field '%s'", name), location);
+
+        auto field = zeek::compat::RecordType_GetFieldType(rtype, idx);
+        std::string fieldName = rtype->FieldName(idx);
+
+        if ( fieldName != name )
+            throw TypeMismatch(hilti::rt::fmt("mismatch in field name: expected '%s', found '%s'", name, fieldName),
+                               location);
+
+        ::zeek::ValPtr v = nullptr;
+
+        if constexpr ( std::is_same<decltype(val), const hilti::rt::Null&>::value ) {
+            // "Null" turns into an unset optional record field.
+        }
+        else
+            // This may return a nullptr in cases where the field is to be left unset.
+            v = to_val(val, field, location);
```

Comment:
Does that include optionals? An unset Spicy-optional should turn into an unset Zeek-side record field. (I know this code is copied from above, but not sure how optionals are handled.)

Reply:
Yes, this seems to largely work for optional fields. I'll update the test for that.

While testing this I also found a bug in how we handle unset `attrs` below where we would have segfaulted on an unset `&optional` Spicy-side field mapped onto a non-`&optional` Zeek-side field. I'll append a patch fixing this here and in the handling of tuples.